src/utils/flatten.js

```diff
@@ -14,28 +14,70 @@ export const objToCss = (obj: Object, prevKey?: string): string => {
 }` : css
 }
 
-const flatten = (chunks: Array<Interpolation>, executionContext: ?Object): Array<Interpolation> => (
-  chunks.reduce((ruleSet: Array<Interpolation>, chunk: ?Interpolation) => {
-    /* Remove falsey values */
-    if (chunk === undefined || chunk === null || chunk === false || chunk === '') return ruleSet
-    /* Flatten ruleSet */
-    if (Array.isArray(chunk)) return [...ruleSet, ...flatten(chunk, executionContext)]
-
-    /* Handle other components */
-    // $FlowFixMe not sure how to make this pass
-    if (chunk.hasOwnProperty('styledComponentId')) return [...ruleSet, `.${chunk.styledComponentId}`]
-
-    /* Either execute or defer the function */
-    if (typeof chunk === 'function') {
-      return executionContext
-        ? ruleSet.concat(...flatten([chunk(executionContext)], executionContext))
-        : ruleSet.concat(chunk)
+const innerFlatten = (
+  targetChunks: Array<Interpolation>,
+  chunk: Interpolation,
+  executionContext: ?Object,
+) => {
+  if (chunk === undefined || chunk === null || chunk === false || chunk === '') {
+    return
+  }
+
+  const chunkType = typeof chunk
+
+  if (chunkType === 'string') {
+    targetChunks.push(chunk)
+    return
+  }
+
+  if (chunkType === 'number') {
+    targetChunks.push(chunk.toString())
+    return
+  }
+
+  if (Array.isArray(chunk)) {
+    for (let i = 0; i < chunk.length; i += 1) {
+      innerFlatten(targetChunks, chunk[i], executionContext)
     }
+    return
+  }
+
+  if (chunkType === 'function') {
+    if (executionContext) {
+      // $FlowFixMe flow thinks chunk is a number
+      innerFlatten(targetChunks, chunk(executionContext), executionContext)
+    } else {
+      // defer the execution context, e.g. `css` styles
+      targetChunks.push(chunk)
+    }
+    return
+  }
+
+  if (chunk.hasOwnProperty('styledComponentId')) {
+    // $FlowFixMe flow thins chunk is a number here
+    targetChunks.push(`.${chunk.styledComponentId}`)
+    return
+  }
 
-    /* Handle objects */
-    // $FlowFixMe have to add %checks somehow to isPlainObject
-    return ruleSet.concat(isPlainObject(chunk) ? objToCss(chunk) : chunk.toString())
-  }, [])
-)
+  if (isPlainObject(chunk)) {
+    // $FlowFixMe flow thinks chunk is a number, but its an object
+    targetChunks.push(objToCss(chunk))
+    return
+  }
+
+
+  // all else fails, just add it, and to string it
+  targetChunks.push(chunk.toString())
+  /* flow-enabled */
+}
+
+const flatten = (chunks: Array<Interpolation>, executionContext: ?Object): Array<Interpolation> => {
+  const resultChunks: Array<Interpolation> = []
+  for (let i = 0; i < chunks.length; i += 1) {
+    innerFlatten(resultChunks, chunks[i], executionContext)
```

Comment:
As mentioned before `reduce` is in real-life contexts almost as fast as an `if` loop. So this is just for "novelty numbers" in benchmarks and won't make it difference compared to what actually is expensive.

Also, I'll need to review this change further, to make sure that this doesn't introduce any regressions